networking/src/Ntp/Util.hs

```diff
@@ -1,97 +1,319 @@
-import           Data.List (find, sortOn)
-import           Data.Time.Clock.POSIX (getPOSIXTime)
-import           Data.Time.Units (Microsecond, fromMicroseconds)
+import           Control.Exception (Exception, IOException, catch, throw)
+import           Control.Monad (void)
+import           Data.Bifunctor (Bifunctor (..))
+import           Data.Binary (encode)
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString.Lazy as LBS
+import           Data.Foldable (traverse_)
+import           Data.List (find)
+import           Data.Semigroup (First (..), Last (..), Option (..),
+                     Semigroup (..))
+import           Data.Text (Text)
+import           Formatting (sformat, shown, (%))
 import           Network.Socket (AddrInfo,
                      AddrInfoFlag (AI_ADDRCONFIG, AI_PASSIVE),
                      Family (AF_INET, AF_INET6), PortNumber (..),
                      SockAddr (..), Socket, SocketOption (ReuseAddr),
                      SocketType (Datagram), aNY_PORT, addrAddress, addrFamily,
-                     addrFlags, addrSocketType, bind, defaultHints,
-                     defaultProtocol, getAddrInfo, setSocketOption, socket)
+                     addrFlags, addrSocketType)
+import qualified Network.Socket as Socket
+import qualified Network.Socket.ByteString as Socket.ByteString (sendTo)
+import qualified System.Wlog as Wlog
+
+import           Ntp.Packet (NtpPacket)
+import           Pos.Util.Trace (Trace, traceWith, wlogTrace)
+
+
+ntpTrace :: Trace IO (Wlog.Severity, Text)
+ntpTrace = wlogTrace "NtpClient"
+
+logWarning :: Text -> IO ()
+logWarning msg = traceWith ntpTrace (Wlog.Warning, msg)
+
+logInfo :: Text -> IO ()
+logInfo msg = traceWith ntpTrace (Wlog.Info, msg)
+
+logDebug :: Text -> IO ()
+logDebug msg = traceWith ntpTrace (Wlog.Debug, msg)
+
+logError :: Text -> IO ()
+logError msg = traceWith ntpTrace (Wlog.Error, msg)
+
+data AddrFamily = IPv4 | IPv6
+    deriving Show
+
+-- |
+-- Newtype wrapper which tags the type with either IPv4 or IPv6 phantom type.
+data WithAddrFamily (t :: AddrFamily) a where
+    WithIPv6 :: a -> WithAddrFamily 'IPv6 a
+    WithIPv4 :: a -> WithAddrFamily 'IPv4 a
+
+instance Show a => Show (WithAddrFamily t a) where
+    show a = show (getAddrFamily a) ++ " " ++ show (runWithAddrFamily a)
+
+instance Eq a => Eq (WithAddrFamily t a) where
+    a == b = runWithAddrFamily a == runWithAddrFamily b
+
+instance Functor (WithAddrFamily t) where
+    fmap f (WithIPv6 a) =  WithIPv6 (f a)
+    fmap f (WithIPv4 a) =  WithIPv4 (f a)
+
+runWithAddrFamily :: WithAddrFamily t a -> a
+runWithAddrFamily (WithIPv6 a) = a
+runWithAddrFamily (WithIPv4 a) = a
+
+getAddrFamily :: WithAddrFamily t a -> AddrFamily
+getAddrFamily (WithIPv6 _) = IPv6
+getAddrFamily (WithIPv4 _) = IPv4
+
+-- |
+-- Keep either of the two types or both.
+data EitherOrBoth a b
+    = EBFirst  !a
+    | EBSecond !b
+    | EBBoth   !a !b
+    deriving (Show, Eq, Ord)
```

Comment:
aka the [`these`](http://hackage.haskell.org/package/these-0.7.4/docs/Data-These.html) datatype, though this is strict

Reply:
Indeed, the strictness doesn't matter that much, since we are just holding max three values, so if you prefer we can pull `these` as a dependency. `these` hasn't been yet used anywhere in `cardano-sl` but it's a small dependency so I don't mind.